### ISSUE TITLE
一个由GLM瞎捣鼓的思路，LoRA rank 通道多样化正则，来解决使用SRA进行快速训练时（10ep），多画风没法好好训练到的问题

### ISSUE THREAD

**Issue by And233** [WalkingMeatAxolotl/AnimaLoraStudio#300]
仅作为draft来作参考，因为我主力训练器不是这里（）

    """
    LoRA rank 通道多样化正则。

    针对多画风训练中的两个核心问题:
    1. 风格混淆: 不同 rank 通道学到相同特征 (通道冗余)
    2. 风格遗忘: 高频风格占据多数通道容量, 低频风格被挤压

    两个正则项, 可独立开关:

    (1) 权重去相关 (lora_diversity_lambda > 0):
        计算 lora_up.weight 的行间 (rank 维) 协方差, 惩罚非对角元。
        鼓励不同 rank 通道学习不同的输出方向 → 缓解通道冗余 → 风格分化。
        数学: corr = B^T @ B (r×r), loss = ||corr - diag(corr)||²_F

    (2) 激活稀疏化 (lora_sparsity_lambda > 0):
        用 forward hook 捕获 lora_down 输出 (rank 维激活 lx),
        惩罚每样本的通道激活熵 → 鼓励每个样本只激活少数 rank 通道。
        不同样本 (风格) 自然走不同通道子集 → 缓解风格遗忘。
        数学: p = |lx|.mean(token_dim) → normalize → entropy(p) → 越低越稀疏

    两个正则都不改变 LoRA 架构和推理行为, 纯训练正则, 可自由组合。
    """

    from __future__ import annotations

    import logging
    from typing import Any, Optional

    import torch
    import torch.nn as nn

    logger = logging.getLogger(__name__)

    class LoRADiversityRegularizer:
        """LoRA rank 通道多样化正则: 权重去相关 + 激活稀疏化。

        用法:
        1. __init__: 传入 network, 读取 lora 模块列表
        2. setup_hooks: 在 lora_down 上注册 forward hook (激活稀疏化用)
        3. compute: 每步调用, 返回总正则 loss
        4. remove_hooks: 训练结束 / 销毁前移除 hook
        """

        def __init__(
            self,
            network: nn.Module,
            diversity_lambda: float = 0.0,
            sparsity_lambda: float = 0.0,
            sparsity_target_count: int = 0,
            sparsity_max_hook_modules: int = 4,
        ):
            """
            Args:
                network: LoRA 网络对象 (LoRANetwork), 含 .unet_loras / .text_encoder_loras
                diversity_lambda: 权重去相关正则强度。0=禁用
                sparsity_lambda: 激活稀疏化正则强度。0=禁用
                sparsity_target_count: 激活稀疏化的目标通道数。
                    0=自动 (lora_dim 的 1/4); >0=期望每样本激活约这么多通道
            """
            self.network = network
            self.diversity_lambda = float(diversity_lambda)
            self.sparsity_lambda = float(sparsity_lambda)
            self.sparsity_target_count = int(sparsity_target_count)
            self.sparsity_max_hook_modules = int(sparsity_max_hook_modules)

            self._hooks: list[Any] = []
            # 激活缓存: {id(lora_module): lx_tensor}
            # 只缓存最后一次 forward 的 lx, compute() 时读取后清空
            self._activation_cache: dict[int, torch.Tensor] = {}

            # 收集所有 LoRA 模块 (统一处理 plain / packed)
            self._lora_modules: list[Any] = self._collect_lora_modules()
            self._lora_dim = self._detect_lora_dim()

            if sparsity_target_count == 0 and self._lora_dim > 0:
                self.sparsity_target_count = max(1, self._lora_dim // 4)

        def _collect_lora_modules(self) -> list[Any]:
            """从 network 收集所有 LoRA 模块 (unet_loras + text_encoder_loras)。"""
            loras = []
            for attr in ("unet_loras", "text_encoder_loras"):
                modules = getattr(self.network, attr, None)
                if modules:
                    loras.extend(modules)
            return loras

        def _detect_lora_dim(self) -> int:
            """检测 lora_dim (rank), 从第一个可用的 lora 模块读取。"""
            for lora in self._lora_modules:
                if hasattr(lora, "lora_dim"):
                    return int(lora.lora_dim)
                # fallback: 从 lora_down.weight 形状推断
                down = getattr(lora, "lora_down", None)
                if isinstance(down, nn.Linear):
                    return int(down.out_features)
                elif isinstance(down, nn.ModuleList) and len(down) > 0:
                    return int(down[0].out_features)
            return 0

        def setup_hooks(self) -> None:
            """在 lora_down 上注册 forward hook, 捕获 rank 维激活 lx。

            ⚠️ 显存关键: hook 持有的激活引用会阻止梯度检查点回收显存。
            故只 hook 少数代表性层 (类比 SRA² 只 hook 一个 block), 而非全部 ~560 个 LoRA。
            均匀采样 unet_loras 的 sparsity_max_hook_modules 个模块。
            """
            if self.sparsity_lambda <= 0.0:
                return

            # 均匀采样: 从 unet_loras 中选 sparsity_max_hook_modules 个代表性模块
            candidates = [l for l in self._lora_modules if getattr(self.network, 'unet_loras', None) and l in self.network.unet_loras]
            if not candidates:
                candidates = self._lora_modules
            n = len(candidates)
            max_hooks = min(self.sparsity_max_hook_modules, n)
            if max_hooks <= 0:
                return
            # 均匀间隔采样 (含首尾)
            if max_hooks == 1:
                sampled = [candidates[n // 2]]
            else:
                indices = [int(i * (n - 1) / (max_hooks - 1)) for i in range(max_hooks)]
                sampled = [candidates[i] for i in indices]

            for lora in sampled:
                down_module = getattr(lora, "lora_down", None)
                if down_module is None:
                    continue
                # packed: lora_down 是 ModuleList, hook 注册在每个 part 上
                targets = down_module if isinstance(down_module, nn.ModuleList) else [down_module]
                lora_id = id(lora)
                for part in targets:
                    if isinstance(part, nn.Module):
                        hook = part.register_forward_hook(self._make_hook(lora_id))
                        self._hooks.append(hook)

            logger.info(f"LoRA 激活稀疏化: hook 注册在 {len(sampled)}/{n} 个代表性模块上 (显存优化)")

        def _make_hook(self, lora_id: int):
            """创建 forward hook 闭包, 捕获 lora_down 输出 lx。"""
            def hook_fn(module, input, output):
                # output: (..., lora_dim) 或 packed 的 (..., parts, lora_dim)
                # detach 不阻断梯度 (hook 在 autograd 图之外, 但 output 本身带梯度)
                self._activation_cache[lora_id] = output
            return hook_fn

        def remove_hooks(self) -> None:
            """移除所有 forward hook。"""
            for h in self._hooks:
                h.remove()
            self._hooks.clear()
            self._activation_cache.clear()

        def compute(self) -> torch.Tensor:
            """计算多样化正则 loss = diversity_lambda · weight_decorrelation
                                    + sparsity_lambda · activation_sparsity

            返回标量 loss (已乘 lambda), 未启用时返回 0。
            """
            total_loss = None

            # ── (1) 权重去相关: 鼓励 rank 通道学习不同方向 ──
            if self.diversity_lambda > 0.0:
                div_loss = self._compute_weight_decorrelation()
                if div_loss is not None:
                    total_loss = self.diversity_lambda * div_loss

            # ── (2) 激活稀疏化: 鼓励每样本只激活少数 rank 通道 ──
            if self.sparsity_lambda > 0.0:
                sparse_loss = self._compute_activation_sparsity()
                if sparse_loss is not None:
                    sp_loss = self.sparsity_lambda * sparse_loss
                    total_loss = sp_loss if total_loss is None else total_loss + sp_loss

            # 清空激活缓存 (每步只读一次)
            self._activation_cache.clear()

            return total_loss if total_loss is not None else torch.tensor(0.0)

        def _get_up_weight(self, lora: Any) -> Optional[torch.Tensor]:
            """获取 lora_up.weight, 处理 packed ModuleList。

            返回拼接后的 (out_dim, lora_dim) 或 None。
            """
            up = getattr(lora, "lora_up", None)
            if up is None:
                return None
            if isinstance(up, nn.ModuleList):
                # packed: 每个 part 是 (out_dim_local, lora_dim), 沿 dim 0 拼接
                weights = [p.weight for p in up if hasattr(p, "weight")]
                if not weights:
                    return None
                return torch.cat(weights, dim=0) if len(weights) > 1 else weights[0]
            elif hasattr(up, "weight"):
                return up.weight
            return None

        def _compute_weight_decorrelation(self) -> Optional[torch.Tensor]:
            """权重去相关: corr = B^T @ B → 惩罚非对角元。

            B = lora_up.weight (out_dim, lora_dim)
            B^T @ B → (lora_dim, lora_dim) 通道间协方差
            非对角元越大 = 通道越相似 → 惩罚。
            """
            total = None
            for lora in self._lora_modules:
                B = self._get_up_weight(lora)
                if B is None or B.dim() < 2:
                    continue
                # B: (out_dim, lora_dim), 需要至少 2 个 rank 通道
                lora_dim = B.shape[-1] if B.shape[0] >= B.shape[1] else B.shape[0]
                if lora_dim < 2:
                    continue

                # 通道间协方差: (lora_dim, lora_dim)
                # lora_up.weight 是 (out_dim, lora_dim), rank 在 dim 1
                corr = B.float().t() @ B.float()  # (lora_dim, lora_dim)

                # 去对角线: 非对角元 = 通道间相似度
                eye = torch.eye(corr.shape[0], device=corr.device, dtype=corr.dtype)
                off_diag = corr * (1.0 - eye)

                # Frobenius 范数归一化 (避免不同层 lora_dim 不同导致尺度差异)
                layer_loss = (off_diag ** 2).sum() / (corr.diagonal() ** 2).sum().clamp(min=1e-12)

                total = layer_loss if total is None else total + layer_loss

            if total is not None and self._lora_modules:
                total = total / len(self._lora_modules)
            return total

        def _compute_activation_sparsity(self) -> Optional[torch.Tensor]:
            """激活稀疏化: 惩罚每样本的 rank 通道激活熵。

            lx 来自 hook 捕获的 lora_down 输出: (..., lora_dim)
            沿非 batch 非 rank 维求均值 → (batch, lora_dim) 激活强度
            归一化为概率分布 → 计算熵 → 越低越稀疏 (只激活少数通道)

            sparsity_target_count 控制期望激活通道数: k
            均匀分布的熵 = log(lora_dim), 完全集中的熵 = 0
            目标熵 ≈ log(k) + (1 - 1/k) ≈ log(k) (k 个等概率通道)
            """
            total = None
            count = 0
            for lora_id, lx in self._activation_cache.items():
                if lx is None:
                    continue
                # lx: (..., lora_dim) 或 packed (..., parts, lora_dim)
                lx_f = lx.float()
                if lx_f.dim() < 2:
                    continue

                # 处理 packed: (..., parts, lora_dim) → 把 parts 合进 lora_dim
                # packed 时 lx 最后一维可能不是 lora_dim 而是别的
                # 安全起见: 假设最后一维是 lora_dim
                rank_dim_size = lx_f.shape[-1]
                if rank_dim_size < 2:
                    continue

                # 沿非 batch 非 rank 维求 |lx| 均值 → (batch, rank_dim)
                # batch 是 dim 0, rank 是 dim -1
                reduce_dims = tuple(range(1, lx_f.dim() - 1))
                if reduce_dims:
                    act = lx_f.abs().mean(dim=reduce_dims)  # (batch, rank_dim)
                else:
                    act = lx_f.abs()  # (batch, rank_dim)

                # 归一化为概率分布
                p = act / (act.sum(dim=-1, keepdim=True) + 1e-12)  # (batch, rank_dim)

                # 熵: H(p) = -sum(p · log(p)), 越低越稀疏
                entropy = -(p * (p + 1e-12).log()).sum(dim=-1)  # (batch,)
                layer_loss = entropy.mean()

                total = layer_loss if total is None else total + layer_loss
                count += 1

            if total is not None and count > 0:
                total = total / count
                # 减去目标熵, 使得当熵接近目标时 loss ≈ 0 (不惩罚已达标的稀疏度)
                # 目标熵 = log(target_count), 用自然对数 (entropy 也是自然对数)
                import math
                target_entropy = math.log(max(self.sparsity_target_count, 1))
                total = (total - target_entropy).clamp(min=0.0)
            return total

        @property
        def is_active(self) -> bool:
            """是否有任何正则项被启用。"""
            return self.diversity_lambda > 0.0 or self.sparsity_lambda > 0.0

        def get_last_stats(self) -> dict[str, Optional[float]]:
            """返回最近的统计信息 (供 logging)。"""
            # 简化: 返回 lambda 值, 真实 loss 值由调用方记录
            return {
                "diversity_lambda": self.diversity_lambda if self.diversity_lambda > 0 else None,
                "sparsity_lambda": self.sparsity_lambda if self.sparsity_lambda > 0 else None,
            }

**Comment by WalkingMeatAxolotl** [comment]
arraneye 好评
